// @looker/components/lib/Drawer/DrawerSurface.js
import styled from 'styled-components';
import { variant } from 'styled-system';
import { SurfaceBase, surfaceTransition } from '../Dialog/SurfaceBase';
import { dialogWidth } from '../Dialog/dialogWidth';
var placement = variant({
  prop: 'placement',
  variants: {
    left: {
      boxShadow: '-18px 0 18px -18px rgba(0, 0, 0, 0.12)',
      left: 0
    },
    right: {
      boxShadow: '-18px 0 18px -18px rgba(0, 0, 0, 0.12)',
      right: 0
    }
  }
});
export var DrawerSurface = styled(SurfaceBase).attrs(function (_ref) {
  var _ref$placement = _ref.placement,
      placement = _ref$placement === void 0 ? 'right' : _ref$placement,
      _ref$width = _ref.width,
      width = _ref$width === void 0 ? 'small' : _ref$width;
  return {
    placement: placement,
    width: width
  };
}).withConfig({
  displayName: "DrawerSurface",
  componentId: "sc-1vwnqjh-0"
})(["height:100%;position:absolute;transition:transform ", ",opacity ", ";", " ", " &.entering,&.exiting{opacity:0.01;transform:translateX( ", " );}"], surfaceTransition, surfaceTransition, placement, dialogWidth, function (_ref2) {
  var placement = _ref2.placement;
  return placement === 'left' ? '-100%' : '100%';
});
