// @looker/design-tokens/lib/color/utils/generateStatefulColors.js
import darken from 'polished/lib/color/darken';
import lighten from 'polished/lib/color/lighten';
import { mixScaledColors } from './mixScaledColors';
export var generateInteractive = function generateInteractive(color) {
  return lighten(0.04, color);
};
export var generatePressed = function generatePressed(color) {
  return darken(0.07, color);
};
export var accentBlendScale = 16;
export var generateStatefulColor = function generateStatefulColor(background, color) {
  return {
    subtle: mixScaledColors(10, color, background),
    accent: mixScaledColors(accentBlendScale, color, background),
    focus: mixScaledColors(60, color, background),
    interactive: generateInteractive(color),
    pressed: generatePressed(color),
    text: background,
    border: color
  };
};
export var generateStatefulColors = function generateStatefulColors(specifiable, derivatives) {
  var background = specifiable.background,
      key = specifiable.key,
      critical = specifiable.critical;
  var neutral = derivatives.neutral;
  var keyColors = generateStatefulColor(background, key);
  var criticalColors = generateStatefulColor(background, critical);
  var neutralColors = generateStatefulColor(background, neutral);
  return {
    keySubtle: keyColors.subtle,
    keyAccent: keyColors.accent,
    keyFocus: keyColors.focus,
    keyInteractive: keyColors.interactive,
    keyPressed: keyColors.pressed,
    keyText: keyColors.text,
    keyBorder: keyColors.border,
    criticalSubtle: criticalColors.subtle,
    criticalAccent: criticalColors.accent,
    criticalFocus: criticalColors.focus,
    criticalInteractive: criticalColors.interactive,
    criticalPressed: criticalColors.pressed,
    criticalText: criticalColors.text,
    criticalBorder: criticalColors.border,
    neutralSubtle: neutralColors.subtle,
    neutralAccent: neutralColors.accent,
    neutralFocus: neutralColors.focus,
    neutralInteractive: neutralColors.interactive,
    neutralPressed: neutralColors.pressed,
    neutralText: neutralColors.text,
    neutralBorder: neutralColors.border
  };
};
