// @looker/components/lib/Panel/PanelHeader.js
import React, { useContext } from 'react';
import { ThemeContext } from 'styled-components';
import { ArrowBack } from '@styled-icons/material-rounded/ArrowBack';
import { ArrowForward } from '@styled-icons/material-rounded/ArrowForward';
import { Heading } from '../Text';
import { Space } from '../Layout';
import { IconButton } from '../Button';
export var PanelHeader = function PanelHeader(_ref) {
  var className = _ref.className,
      direction = _ref.direction,
      handleClose = _ref.handleClose,
      title = _ref.title;

  var _useContext = useContext(ThemeContext),
      space = _useContext.space;

  return React.createElement(Space, {
    as: "header",
    className: className,
    height: space.xxlarge,
    px: "large",
    gap: "small",
    mt: "small",
    mb: "1.5rem"
  }, React.createElement(IconButton, {
    icon: direction === 'left' ? React.createElement(ArrowBack, null) : React.createElement(ArrowForward, null),
    label: "Close ".concat(title),
    onClick: handleClose,
    size: "small"
  }), React.createElement(Heading, {
    fontSize: "xlarge"
  }, title));
};
