// @looker/components/lib/Panel/PanelSurface.js
import "core-js/modules/es.array.concat.js";
import styled, { css } from 'styled-components';
import { variant } from 'styled-system';

var surfaceTransition = function surfaceTransition() {
  return css(["", ""], function (_ref) {
    var theme = _ref.theme;
    return "".concat(theme.transitions.moderate, "ms ").concat(theme.easings.ease);
  });
};

var direction = variant({
  prop: 'direction',
  variants: {
    left: {
      transform: 'translate(-100%, 0);'
    },
    right: {
      transform: 'translate(100%, 0);'
    }
  }
});
export var PanelSurface = styled.div.attrs(function (_ref2) {
  var _ref2$direction = _ref2.direction,
      direction = _ref2$direction === void 0 ? 'left' : _ref2$direction;
  return {
    direction: direction
  };
}).withConfig({
  displayName: "PanelSurface",
  componentId: "sc-1ptnzaf-0"
})(["background:", ";bottom:0;height:100%;left:0;position:absolute;right:0;top:0;transition:transform ", ",opacity ", ";width:100%;&.entering,&.exiting{", "}"], function (_ref3) {
  var theme = _ref3.theme;
  return theme.colors.background;
}, surfaceTransition, surfaceTransition, direction);
