// @looker/components/lib/cjs/Form/Fields/FieldInline.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FieldInline = void 0;

require("core-js/modules/es.symbol.description.js");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Label = require("../Label/Label");

var _Text = require("../../Text");

var _ValidationMessage = require("../ValidationMessage/ValidationMessage");

var _Truncate = require("../../Truncate");

var _RequiredStar = require("./RequiredStar");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const FieldInlineLayout = ({
  className,
  children,
  description,
  detail,
  label,
  id,
  required,
  validationMessage
}) => {
  return _react.default.createElement("label", {
    className: className,
    htmlFor: id
  }, detail && _react.default.createElement(FieldDetail, null, detail), _react.default.createElement(InputArea, null, children), _react.default.createElement(_Label.Label, {
    as: "span"
  }, _react.default.createElement(_Truncate.Truncate, null, label), required && _react.default.createElement(_RequiredStar.RequiredStar, null)), _react.default.createElement(MessageArea, {
    id: `${id}-describedby`
  }, description && _react.default.createElement(FieldDescription, null, description), validationMessage ? _react.default.createElement(_ValidationMessage.ValidationMessage, validationMessage) : null));
};

const FieldDetail = (0, _styledComponents.default)(_Text.Paragraph).withConfig({
  displayName: "FieldInline__FieldDetail",
  componentId: "kl4qjf-0"
})(["color:", ";font-size:", ";grid-column:3;grid-row:1;-ms-grid-column:3;-ms-grid-column-span:2;-ms-grid-row:1;padding-left:", ";"], ({
  theme
}) => theme.colors.text2, ({
  theme
}) => theme.fontSizes.xsmall, ({
  theme
}) => theme.space.xsmall);
const FieldDescription = (0, _styledComponents.default)(_Text.Paragraph).withConfig({
  displayName: "FieldInline__FieldDescription",
  componentId: "kl4qjf-1"
})(["color:", ";font-size:", ";"], ({
  theme
}) => theme.colors.text2, ({
  theme
}) => theme.fontSizes.xsmall);

const InputArea = _styledComponents.default.div.withConfig({
  displayName: "FieldInline__InputArea",
  componentId: "kl4qjf-2"
})(["grid-column:1;grid-row:1;padding-right:", ";-ms-grid-column:1;-ms-grid-column-span:1;-ms-grid-row:1;"], ({
  theme: {
    space
  }
}) => space.xsmall);

const MessageArea = _styledComponents.default.div.withConfig({
  displayName: "FieldInline__MessageArea",
  componentId: "kl4qjf-3"
})(["grid-column:2;grid-column-end:span 2;grid-row:2;-ms-grid-column:2;-ms-grid-column-end:span 2;-ms-grid-column-span:2;-ms-grid-row:2;"]);

const FieldInline = (0, _styledComponents.default)(FieldInlineLayout).withConfig({
  displayName: "FieldInline",
  componentId: "kl4qjf-4"
})(["align-items:center;display:grid;grid-template-columns:auto auto auto auto;justify-content:start;line-height:", ";display:-ms-grid;-ms-grid-columns:auto auto auto auto;", "{align-items:center;color:", ";display:flex;font-size:", ";font-weight:normal;grid-column:2;grid-row:1;width:fit-content;-ms-grid-column:2;-ms-grid-column-span:1;-ms-grid-row:1;}"], ({
  theme
}) => theme.lineHeights.small, _Label.Label, ({
  theme,
  disabled
}) => disabled && theme.colors.text1, ({
  theme
}) => theme.fontSizes.small);
exports.FieldInline = FieldInline;
