// @looker/components/lib/Form/Fields/FieldInline.js
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.symbol.description.js";
import "core-js/modules/es.string.small.js";
import React from 'react';
import styled from 'styled-components';
import { Label } from '../Label/Label';
import { Paragraph } from '../../Text';
import { ValidationMessage } from '../ValidationMessage/ValidationMessage';
import { Truncate } from '../../Truncate';
import { RequiredStar } from './RequiredStar';

var FieldInlineLayout = function FieldInlineLayout(_ref) {
  var className = _ref.className,
      children = _ref.children,
      description = _ref.description,
      detail = _ref.detail,
      label = _ref.label,
      id = _ref.id,
      required = _ref.required,
      validationMessage = _ref.validationMessage;
  return React.createElement("label", {
    className: className,
    htmlFor: id
  }, detail && React.createElement(FieldDetail, null, detail), React.createElement(InputArea, null, children), React.createElement(Label, {
    as: "span"
  }, React.createElement(Truncate, null, label), required && React.createElement(RequiredStar, null)), React.createElement(MessageArea, {
    id: "".concat(id, "-describedby")
  }, description && React.createElement(FieldDescription, null, description), validationMessage ? React.createElement(ValidationMessage, validationMessage) : null));
};

var FieldDetail = styled(Paragraph).withConfig({
  displayName: "FieldInline__FieldDetail",
  componentId: "kl4qjf-0"
})(["color:", ";font-size:", ";grid-column:3;grid-row:1;-ms-grid-column:3;-ms-grid-column-span:2;-ms-grid-row:1;padding-left:", ";"], function (_ref2) {
  var theme = _ref2.theme;
  return theme.colors.text2;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.fontSizes.xsmall;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.space.xsmall;
});
var FieldDescription = styled(Paragraph).withConfig({
  displayName: "FieldInline__FieldDescription",
  componentId: "kl4qjf-1"
})(["color:", ";font-size:", ";"], function (_ref5) {
  var theme = _ref5.theme;
  return theme.colors.text2;
}, function (_ref6) {
  var theme = _ref6.theme;
  return theme.fontSizes.xsmall;
});
var InputArea = styled.div.withConfig({
  displayName: "FieldInline__InputArea",
  componentId: "kl4qjf-2"
})(["grid-column:1;grid-row:1;padding-right:", ";-ms-grid-column:1;-ms-grid-column-span:1;-ms-grid-row:1;"], function (_ref7) {
  var space = _ref7.theme.space;
  return space.xsmall;
});
var MessageArea = styled.div.withConfig({
  displayName: "FieldInline__MessageArea",
  componentId: "kl4qjf-3"
})(["grid-column:2;grid-column-end:span 2;grid-row:2;-ms-grid-column:2;-ms-grid-column-end:span 2;-ms-grid-column-span:2;-ms-grid-row:2;"]);
export var FieldInline = styled(FieldInlineLayout).withConfig({
  displayName: "FieldInline",
  componentId: "kl4qjf-4"
})(["align-items:center;display:grid;grid-template-columns:auto auto auto auto;justify-content:start;line-height:", ";display:-ms-grid;-ms-grid-columns:auto auto auto auto;", "{align-items:center;color:", ";display:flex;font-size:", ";font-weight:normal;grid-column:2;grid-row:1;width:fit-content;-ms-grid-column:2;-ms-grid-column-span:1;-ms-grid-row:1;}"], function (_ref8) {
  var theme = _ref8.theme;
  return theme.lineHeights.small;
}, Label, function (_ref9) {
  var theme = _ref9.theme,
      disabled = _ref9.disabled;
  return disabled && theme.colors.text1;
}, function (_ref10) {
  var theme = _ref10.theme;
  return theme.fontSizes.small;
});
