// @looker/components/lib/Calendar/CalendarNav.js
import "core-js/modules/es.date.to-string.js";
import "core-js/modules/es.date.now.js";
import { useTranslation } from 'react-i18next';
import React, { useContext } from 'react';
import styled from 'styled-components';
import noop from 'lodash/noop';
import { ChevronLeft } from '@styled-icons/material-rounded/ChevronLeft';
import { ChevronRight } from '@styled-icons/material-rounded/ChevronRight';
import { Tooltip } from '../Tooltip';
import { IconButton, ButtonTransparent } from '../Button';
import { Heading } from '../Text';
import { CalendarContext } from './CalendarContext';
export var headingSizeMap = function headingSizeMap(size) {
  switch (size) {
    case 'small':
      return 'h6';

    case 'medium':
      return 'h5';

    case 'large':
      return 'h4';

    default:
      return 'h5';
  }
};
export var CalendarNav = function CalendarNav(_ref) {
  var month = _ref.month,
      labels = _ref.labels,
      localeUtils = _ref.localeUtils,
      nextMonth = _ref.nextMonth,
      previousMonth = _ref.previousMonth;

  var _useTranslation = useTranslation('CalendarNav'),
      t = _useTranslation.t;

  var _useContext = useContext(CalendarContext),
      size = _useContext.size,
      _useContext$onNowClic = _useContext.onNowClick,
      onNowClick = _useContext$onNowClic === void 0 ? noop : _useContext$onNowClic,
      _useContext$onNextCli = _useContext.onNextClick,
      onNextClick = _useContext$onNextCli === void 0 ? noop : _useContext$onNextCli,
      _useContext$onPrevCli = _useContext.onPrevClick,
      onPrevClick = _useContext$onPrevCli === void 0 ? noop : _useContext$onPrevCli,
      showPreviousButton = _useContext.showPreviousButton,
      showNextButton = _useContext.showNextButton;

  var handleNextClick = function handleNextClick(e) {
    e.preventDefault();
    onNextClick(nextMonth);
  };

  var handlePreviousClick = function handlePreviousClick(e) {
    e.preventDefault();
    onPrevClick(previousMonth);
  };

  var handleLabelClick = function handleLabelClick() {
    onNowClick(new Date(Date.now()));
  };

  return React.createElement(NavGrid, null, React.createElement(NextButtonWrapper, null, showPreviousButton && React.createElement(IconButton, {
    icon: React.createElement(ChevronLeft, null),
    label: labels.previousMonth,
    size: size,
    onClick: handlePreviousClick
  })), React.createElement(Tooltip, {
    content: t('View Current Month')
  }, React.createElement(ButtonTransparent, {
    onClick: handleLabelClick,
    color: "neutral"
  }, React.createElement(Heading, {
    as: headingSizeMap(size),
    fontWeight: "semiBold",
    fontFamily: "body"
  }, localeUtils.formatMonthTitle(month)))), React.createElement(PrevButtonWrapper, null, showNextButton && React.createElement(IconButton, {
    icon: React.createElement(ChevronRight, null),
    label: labels.nextMonth,
    size: size,
    onClick: handleNextClick,
    style: {
      justifySelf: 'right'
    }
  })));
};
var NavGrid = styled.div.withConfig({
  displayName: "CalendarNav__NavGrid",
  componentId: "sc-4c8g0d-0"
})(["align-items:center;display:grid;grid-template-columns:40px 1fr 40px;justify-items:center;"]);
var NextButtonWrapper = styled.div.withConfig({
  displayName: "CalendarNav__NextButtonWrapper",
  componentId: "sc-4c8g0d-1"
})(["justify-self:left;"]);
var PrevButtonWrapper = styled.div.withConfig({
  displayName: "CalendarNav__PrevButtonWrapper",
  componentId: "sc-4c8g0d-2"
})(["justify-self:right;"]);
