// @looker/components/lib/cjs/Calendar/calendar-size.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.calendarSpacing = exports.calendarSize = exports.daySizeMap = void 0;

var _styledSystem = require("styled-system");

const daySizeMap = {
  small: 32,
  medium: 36,
  large: 44
};
exports.daySizeMap = daySizeMap;
const calendarSize = (0, _styledSystem.variant)({
  prop: 'size',
  variants: {
    small: {
      fontSize: 'xsmall',
      height: `${daySizeMap.small}px`,
      width: `${daySizeMap.small}px`
    },
    medium: {
      fontSize: 'small',
      height: `${daySizeMap.medium}px`,
      width: `${daySizeMap.medium}px`
    },
    large: {
      fontSize: 'medium',
      height: `${daySizeMap.large}px`,
      width: `${daySizeMap.large}px`
    }
  }
});
exports.calendarSize = calendarSize;
const calendarSpacing = (0, _styledSystem.variant)({
  prop: 'size',
  variants: {
    small: {
      p: 'small'
    },
    medium: {
      p: 'medium'
    },
    large: {
      p: 'large'
    }
  }
});
exports.calendarSpacing = calendarSpacing;
