// @looker/components/lib/cjs/Calendar/Calendar.js
"use strict";

require("core-js/modules/es.array.filter.js");

require("core-js/modules/es.object.get-own-property-descriptors.js");

require("core-js/modules/es.weak-map.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Calendar = void 0;

require("core-js/modules/es.array.iterator.js");

var _react = _interopRequireDefault(require("react"));

var _reactDayPicker = _interopRequireWildcard(require("react-day-picker"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _has = _interopRequireDefault(require("lodash/has"));

var _noop = _interopRequireDefault(require("lodash/noop"));

var _designTokens = require("@looker/design-tokens");

var _InputText = require("../Form/Inputs/InputText");

var _calendarSize = require("./calendar-size");

var _CalendarContext = require("./CalendarContext");

var _CalendarNav = require("./CalendarNav");

var _dayPickerCss = require("./dayPickerCss");

var _CalendarNavDisabled = require("./CalendarNavDisabled");

var _formatMonthTitle = require("./formatMonthTitle");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const NoopComponent = () => null;

const InternalCalendar = ({
  className,
  disabled,
  localization,
  onDayClick,
  onMonthChange,
  onNextClick,
  onNowClick,
  onPrevClick,
  readOnly,
  selectedDates,
  showNextButton = true,
  showPreviousButton = true,
  viewMonth
}) => {
  const renderDateRange = selectedDates && (0, _has.default)(selectedDates, 'from');
  const modifiers = renderDateRange ? selectedDates : {};

  const disableCallback = (cb = _noop.default) => {
    return (...args) => disabled ? (0, _noop.default)() : cb(...args);
  };

  return _react.default.createElement(_CalendarContext.CalendarContext.Provider, {
    value: {
      onNextClick: disableCallback(onNextClick),
      onNowClick: disableCallback(onNowClick),
      onPrevClick: disableCallback(onPrevClick),
      showNextButton: (!disabled || !readOnly) && showNextButton,
      showPreviousButton: (!disabled || !readOnly) && showPreviousButton,
      size: 'medium'
    }
  }, _react.default.createElement(_reactDayPicker.default, _extends({}, localization || {}, {
    localeUtils: _objectSpread(_objectSpread({}, _reactDayPicker.LocaleUtils), {}, {
      formatMonthTitle: (0, _formatMonthTitle.formatMonthTitle)(localization)
    }),
    className: `${renderDateRange && 'render-date-range'} ${className}`,
    selectedDays: selectedDates,
    month: viewMonth,
    showOutsideDays: true,
    onDayClick: disableCallback(onDayClick),
    navbarElement: disabled ? _CalendarNavDisabled.CalendarNavDisabled : _CalendarNav.CalendarNav,
    captionElement: NoopComponent,
    modifiers: modifiers,
    onMonthChange: onMonthChange
  })));
};

const Calendar = (0, _styledComponents.default)(InternalCalendar).attrs(() => ({
  size: 'medium'
})).withConfig({
  displayName: "Calendar",
  componentId: "sc-16c71io-0"
})(["", " ", " ", " .DayPicker-wrapper{border:1px solid transparent;padding:0;&:focus{outline:none;", "}}.DayPicker-Month{margin:0;padding:0;}.DayPicker-Body{display:grid;grid-gap:1px;}.DayPicker-Week,.DayPicker-WeekdaysRow{display:grid;grid-template-columns:repeat(7,auto);}.DayPicker-Day{", " align-items:center;border:1px solid transparent;color:", ";cursor:", ";display:grid;justify-items:center;line-height:1;padding:0;transition:background-color 110ms linear;&.DayPicker-Day--outside{color:", ";}&--today{color:", ";}&--selected:not(.DayPicker-Day--disabled):not(.DayPicker-Day--outside){background-color:", ";position:static;&:hover{background-color:", ";}}&:focus{border-color:", ";border-width:2px;outline:none;}}&.render-date-range{.DayPicker-Day--selected{border-radius:0;&:not(.DayPicker-Day--from):first-child{border-bottom-left-radius:", ";border-top-left-radius:", ";}&:not(.DayPicker-Day--to):last-child{border-bottom-right-radius:", ";border-top-right-radius:", ";}&.DayPicker-Day--outside,&:not(.DayPicker-Day--to):not(.DayPicker-Day--from){background-color:", ";", "}&:not(.DayPicker-Day--to):not(.DayPicker-Day--from):not(.DayPicker-Day--outside){color:", ";}}.DayPicker-Day--from{border-bottom-left-radius:50%;border-top-left-radius:50%;}.DayPicker-Day--to{border-bottom-right-radius:50%;border-top-right-radius:50%;}}&:not(.DayPicker--interactionDisabled){.DayPicker-Day:not(.DayPicker-Day--disabled):not(.DayPicker-Day--selected):not(.DayPicker-Day--outside):hover{&:hover{background-color:", ";color:", ";}}}"], _designTokens.reset, _dayPickerCss.dayPickerCss, _calendarSize.calendarSpacing, ({
  disabled
}) => !disabled && _InputText.inputTextFocus, _calendarSize.calendarSize, ({
  theme: {
    colors
  },
  disabled
}) => disabled ? colors.text2 : colors.text4, ({
  disabled
}) => disabled ? 'default' : 'pointer', ({
  theme: {
    colors
  }
}) => colors.text1, ({
  theme: {
    colors
  },
  disabled
}) => !disabled && colors.key, ({
  theme: {
    colors
  },
  disabled
}) => disabled ? colors.neutral : colors.key, ({
  theme: {
    colors
  },
  disabled,
  readOnly
}) => disabled || readOnly ? colors.neutralInteractive : colors.keyInteractive, ({
  theme: {
    colors
  },
  disabled,
  readOnly
}) => disabled || readOnly ? 'transparent' : colors.keyBorder, ({
  theme: {
    radii
  }
}) => radii.medium, ({
  theme: {
    radii
  }
}) => radii.medium, ({
  theme: {
    radii
  }
}) => radii.medium, ({
  theme: {
    radii
  }
}) => radii.medium, ({
  theme: {
    colors
  },
  disabled
}) => disabled ? colors.neutralAccent : colors.keyAccent, _designTokens.calendarMixColor, ({
  theme: {
    colors
  }
}) => colors.neutralInteractive, ({
  theme: {
    colors
  },
  disabled
}) => disabled ? 'transparent' : colors.keyAccent, ({
  theme: {
    colors
  },
  disabled
}) => !disabled && colors.key);
exports.Calendar = Calendar;
