// @looker/design-tokens/lib/utils/helpers.js
import rgba from 'polished/lib/color/rgba';
import lighten from 'polished/lib/color/lighten';
import mix from 'polished/lib/color/mix';
import { css } from 'styled-components';
export var buttonShadow = function buttonShadow() {
  var color = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'key';
  return css(["box-shadow:0 0 0 0.15rem ", ";"], function (_ref) {
    var theme = _ref.theme;
    return rgba(theme.colors[color], 0.25);
  });
};
export var iconButtonColorDerivation = function iconButtonColorDerivation() {
  return css(["color:", ";"], function (_ref2) {
    var theme = _ref2.theme;
    return lighten(0.14, theme.colors.neutral);
  });
};
export var tabShadowColor = function tabShadowColor() {
  return css(["box-shadow:0 0 0 0.15rem ", ";"], function (_ref3) {
    var theme = _ref3.theme;
    return rgba(theme.colors.keyFocus, 0.25);
  });
};
export var calendarMixColor = function calendarMixColor() {
  return css(["color:", ";"], function (_ref4) {
    var colors = _ref4.theme.colors;
    return mix(0.65, colors.keyAccent, colors.neutralInteractive);
  });
};
export var knobShadowColor = function knobShadowColor() {
  return css(["box-shadow:0 0 0.01rem 0.01rem ", ";"], function (_ref5) {
    var theme = _ref5.theme;
    return rgba(theme.colors.keyInteractive, 0.5);
  });
};
export var toggleSwitchShadowColor = function toggleSwitchShadowColor() {
  return css(["box-shadow:0 0 0 0.2rem ", ";"], function (_ref6) {
    var theme = _ref6.theme;
    return rgba(theme.colors.keyInteractive, 0.4);
  });
};
export var disabledSwatchColor = function disabledSwatchColor(color) {
  return color && color !== 'transparent' ? rgba(color, 0.85) : undefined;
};
