// @looker/components/lib/Form/Fields/FieldDateRange/FieldDateRange.js
import "core-js/modules/es.object.assign.js";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { useID } from '../../../utils';
import { useFormContext } from '../../Form';
import { InputDateRange } from '../../Inputs/InputDateRange';
import { Field, omitFieldProps, pickFieldProps } from '../Field';
var FieldDateRangeComponent = forwardRef(function (props, ref) {
  var validationMessage = useFormContext(props);
  var id = useID(props.id);
  return React.createElement(Field, _extends({}, pickFieldProps(props), {
    id: id,
    validationMessage: validationMessage
  }), React.createElement(InputDateRange, _extends({}, omitFieldProps(props), {
    "aria-describedby": "describedby-".concat(id),
    "aria-labelledby": "labelledby-".concat(id),
    id: id,
    validationType: validationMessage && validationMessage.type,
    ref: ref
  })));
});
FieldDateRangeComponent.displayName = 'FieldDateRangeComponent';
export var FieldDateRange = styled(FieldDateRangeComponent).withConfig({
  displayName: "FieldDateRange",
  componentId: "xojxrj-0"
})([""]);
