// @looker/design-tokens/lib/cjs/system/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  border: true,
  boxShadow: true,
  color: true,
  display: true,
  flexbox: true,
  fontSize: true,
  height: true,
  layout: true,
  overflow: true,
  padding: true,
  position: true,
  space: true,
  size: true,
  typography: true,
  verticalAlign: true,
  width: true,
  userSelect: true,
  cursor: true,
  Easings: true,
  RadiusSizes: true,
  Radii: true,
  Shadows: true
};
Object.defineProperty(exports, "border", {
  enumerable: true,
  get: function get() {
    return _styledSystem.border;
  }
});
Object.defineProperty(exports, "boxShadow", {
  enumerable: true,
  get: function get() {
    return _styledSystem.boxShadow;
  }
});
Object.defineProperty(exports, "color", {
  enumerable: true,
  get: function get() {
    return _styledSystem.color;
  }
});
Object.defineProperty(exports, "display", {
  enumerable: true,
  get: function get() {
    return _styledSystem.display;
  }
});
Object.defineProperty(exports, "flexbox", {
  enumerable: true,
  get: function get() {
    return _styledSystem.flexbox;
  }
});
Object.defineProperty(exports, "fontSize", {
  enumerable: true,
  get: function get() {
    return _styledSystem.fontSize;
  }
});
Object.defineProperty(exports, "height", {
  enumerable: true,
  get: function get() {
    return _styledSystem.height;
  }
});
Object.defineProperty(exports, "layout", {
  enumerable: true,
  get: function get() {
    return _styledSystem.layout;
  }
});
Object.defineProperty(exports, "overflow", {
  enumerable: true,
  get: function get() {
    return _styledSystem.overflow;
  }
});
Object.defineProperty(exports, "padding", {
  enumerable: true,
  get: function get() {
    return _styledSystem.padding;
  }
});
Object.defineProperty(exports, "position", {
  enumerable: true,
  get: function get() {
    return _styledSystem.position;
  }
});
Object.defineProperty(exports, "space", {
  enumerable: true,
  get: function get() {
    return _styledSystem.space;
  }
});
Object.defineProperty(exports, "size", {
  enumerable: true,
  get: function get() {
    return _styledSystem.size;
  }
});
Object.defineProperty(exports, "typography", {
  enumerable: true,
  get: function get() {
    return _styledSystem.typography;
  }
});
Object.defineProperty(exports, "verticalAlign", {
  enumerable: true,
  get: function get() {
    return _styledSystem.verticalAlign;
  }
});
Object.defineProperty(exports, "width", {
  enumerable: true,
  get: function get() {
    return _styledSystem.width;
  }
});
Object.defineProperty(exports, "userSelect", {
  enumerable: true,
  get: function get() {
    return _userSelect.userSelect;
  }
});
Object.defineProperty(exports, "cursor", {
  enumerable: true,
  get: function get() {
    return _cursor.cursor;
  }
});
Object.defineProperty(exports, "Easings", {
  enumerable: true,
  get: function get() {
    return _easings.Easings;
  }
});
Object.defineProperty(exports, "RadiusSizes", {
  enumerable: true,
  get: function get() {
    return _radii.RadiusSizes;
  }
});
Object.defineProperty(exports, "Radii", {
  enumerable: true,
  get: function get() {
    return _radii.Radii;
  }
});
Object.defineProperty(exports, "Shadows", {
  enumerable: true,
  get: function get() {
    return _shadows.Shadows;
  }
});

var _styledSystem = require("styled-system");

var _userSelect = require("./userSelect");

var _cursor = require("./cursor");

var _easings = require("./easings");

var _radii = require("./radii");

var _shadows = require("./shadows");

var _transitions = require("./transitions");

Object.keys(_transitions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _transitions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _transitions[key];
    }
  });
});

var _size = require("./size");

Object.keys(_size).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _size[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _size[key];
    }
  });
});

var _space = require("./space");

Object.keys(_space).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _space[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _space[key];
    }
  });
});

var _typography = require("./typography");

Object.keys(_typography).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _typography[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _typography[key];
    }
  });
});

var _reset = require("./reset");

Object.keys(_reset).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reset[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _reset[key];
    }
  });
});
