// @looker/components/lib/List/ListItemLayout.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import "core-js/modules/es.symbol.js";
import "core-js/modules/es.symbol.description.js";
import "core-js/modules/es.object.define-property.js";
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/es.array.for-each.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import "core-js/modules/es.object.define-properties.js";
import React from 'react';
import styled, { css } from 'styled-components';
import { color as colorHelper } from '@looker/design-tokens';
import { StyledIconBase } from '@styled-icons/styled-icon';
import { IconPlaceholder } from '../Icon';
import { Flex } from '../Layout';
import { listItemPadding } from './utils';

var ListItemLayoutInternal = function ListItemLayoutInternal(_ref) {
  var children = _ref.children,
      _ref$className = _ref.className,
      className = _ref$className === void 0 ? '' : _ref$className,
      labelCreator = _ref.labelCreator,
      description = _ref.description,
      detail = _ref.detail,
      icon = _ref.icon;
  var content = React.createElement(React.Fragment, null, icon, React.createElement(Flex, {
    flexDirection: "column",
    minWidth: 0,
    flexGrow: 1
  }, children, description), detail);
  return labelCreator({
    children: content,
    className: className
  });
};

export var listItemIconCSS = css(["& > svg,& > ", ",& > ", "{", " align-self:center;flex-grow:0;flex-shrink:0;height:", ";margin-right:", ";width:", ";}"], StyledIconBase, IconPlaceholder, colorHelper, function (_ref2) {
  var iconSize = _ref2.iconSize,
      theme = _ref2.theme;
  return theme.sizes[iconSize];
}, function (_ref3) {
  var iconGap = _ref3.iconGap,
      theme = _ref3.theme;
  return theme.space[iconGap];
}, function (_ref4) {
  var iconSize = _ref4.iconSize,
      theme = _ref4.theme;
  return theme.sizes[iconSize];
});
export var ListItemLayout = styled(ListItemLayoutInternal).attrs(function (_ref5) {
  var _ref5$color = _ref5.color,
      color = _ref5$color === void 0 ? 'text2' : _ref5$color,
      disabled = _ref5.disabled;
  return {
    color: disabled ? 'text1' : color
  };
}).withConfig({
  displayName: "ListItemLayout",
  componentId: "sc-1kn6g5n-0"
})(["", " ", ""], function (props) {
  return listItemPadding(_objectSpread({}, props));
}, listItemIconCSS);
