// @looker/components/lib/cjs/Dialog/DialogSurface.js
"use strict";

require("core-js/modules/es.weak-map.js");

require("core-js/modules/es.array.iterator.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DialogSurface = exports.dialogPlacements = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _styledSystem = require("styled-system");

var _designTokens = require("@looker/design-tokens");

var _SurfaceBase = require("../Dialog/SurfaceBase");

var _dialogWidth = require("./dialogWidth");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const dialogPlacements = ['center', 'cover', 'top'];
exports.dialogPlacements = dialogPlacements;
const space = _designTokens.theme.space,
      breakpoints = _designTokens.theme.breakpoints;
const gapSpace = 'xxlarge';
const coverDimension = `calc(100% - ${space[gapSpace]} * 2)`;
const placements = {
  center: (0, _styledComponents.css)(["align-self:flex-start;max-height:100%;@media screen and (min-width:", "){align-self:center;max-height:", ";}"], breakpoints[0], coverDimension),
  cover: (0, _styledComponents.css)(["height:100%;@media screen and (min-width:", "){height:", ";width:", ";}@media screen and (min-width:", "){height:", ";width:", ";}"], breakpoints[0], coverDimension, coverDimension, breakpoints[1], coverDimension, coverDimension),
  top: (0, _styledComponents.css)(["align-self:flex-start;margin-top:0;max-height:100%;@media screen and (min-width:", "){margin-top:", ";max-height:", ";}"], breakpoints[0], ({
    theme
  }) => theme.space[gapSpace], coverDimension)
};
const defaultDialogSurfacePlacement = 'center';
const DialogSurface = (0, _styledComponents.default)(_SurfaceBase.SurfaceBase).attrs(({
  placement = defaultDialogSurfacePlacement,
  width = 'medium'
}) => ({
  placement,
  width
})).withConfig({
  displayName: "DialogSurface",
  componentId: "sg86rj-0"
})(["box-shadow:", ";position:relative;transition:transform ", ",opacity ", ";", " ", " ", " @media screen and (min-width:", "){border-radius:", ";}&.entering,&.exiting{opacity:0.01;transform:translateY(100%);}"], ({
  theme
}) => theme.shadows[5], _SurfaceBase.surfaceTransition, _SurfaceBase.surfaceTransition, _dialogWidth.dialogWidth, ({
  placement
}) => placements[placement || defaultDialogSurfacePlacement], _styledSystem.height, breakpoints[0], ({
  theme
}) => theme.radii.medium);
exports.DialogSurface = DialogSurface;
