// @looker/components/lib/Portal/Portal.js
import "core-js/modules/es.object.assign.js";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useEffect, useRef } from 'react';
import { createPortal } from 'react-dom';
import styled from 'styled-components';
var rootId = 'modal-root';
export var getPortalRoot = function getPortalRoot() {
  var existing = document.getElementById(rootId);

  if (existing) {
    return existing;
  } else {
    var newElement = document.createElement('div');
    newElement.id = 'modal-root';
    document.body.appendChild(newElement);
    return newElement;
  }
};
export var Portal = forwardRef(function (props, ref) {
  var el = useRef(document.createElement('div'));
  useEffect(function () {
    var root = getPortalRoot();
    if (!root) return;
    var elCurrent = el.current;
    root.appendChild(elCurrent);
    return function () {
      root.removeChild(elCurrent);
    };
  }, [el]);
  var content = React.createElement(InvisiBox, _extends({
    ref: ref
  }, props));
  return createPortal(content, el.current);
});
Portal.displayName = 'Portal';
var InvisiBox = styled.div.withConfig({
  displayName: "Portal__InvisiBox",
  componentId: "sc-8jnv99-0"
})(["align-items:", ";bottom:0;display:flex;justify-content:", ";left:0;pointer-events:none;position:fixed;right:0;top:0;z-index:", ";> *{pointer-events:auto;}"], function (_ref) {
  var vertical = _ref.vertical;
  return vertical === 'top' ? 'flex-start' : vertical === 'bottom' ? 'flex-end' : 'center';
}, function (_ref2) {
  var horizontal = _ref2.horizontal;
  return horizontal === 'left' ? 'flex-start' : horizontal === 'right' ? 'flex-end' : 'center';
}, function (_ref3) {
  var zIndexFloor = _ref3.theme.zIndexFloor;
  return zIndexFloor;
});
