// @looker/components/lib/Form/Inputs/Radio/Radio.js
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.array.index-of.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.object.assign.js";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { reset, space } from '@looker/design-tokens';
import { pickInputProps } from '../InputProps';
import { FauxRadio } from './FauxRadio';
var RadioLayout = forwardRef(function (props, ref) {
  var className = props.className,
      validationType = props.validationType,
      restProps = _objectWithoutProperties(props, ["className", "validationType"]);

  return React.createElement("div", {
    className: className
  }, React.createElement("input", _extends({}, pickInputProps(restProps), {
    "aria-invalid": validationType === 'error' ? 'true' : undefined,
    ref: ref,
    type: "radio"
  })), React.createElement(FauxRadio, null));
});
RadioLayout.displayName = 'RadioLayout';
export var Radio = styled(RadioLayout).withConfig({
  displayName: "Radio",
  componentId: "sc-1e6vir3-0"
})(["", " ", " height:1rem;position:relative;width:1rem;input{background:", ";height:100%;opacity:0;position:absolute;width:100%;z-index:1;&:disabled{cursor:not-allowed;}}input + ", "{border-color:", ";}input:checked + ", "{color:", ";}input:not(:checked) + ", "{background:", ";}input:focus + ", "{border-color:", ";box-shadow:0 0 0 2px ", ";outline:none;}input:disabled + ", "{color:", ";}input:disabled:not(:checked) + ", "{background:", ";color:transparent;&::after{background:", ";}}"], reset, space, function (props) {
  return props.theme.colors.field;
}, FauxRadio, function (_ref) {
  var theme = _ref.theme,
      validationType = _ref.validationType;
  return validationType === 'error' ? theme.colors.criticalBorder : theme.colors.ui2;
}, FauxRadio, function (_ref2) {
  var theme = _ref2.theme;
  return theme.colors.key;
}, FauxRadio, function (_ref3) {
  var theme = _ref3.theme;
  return theme.colors.field;
}, FauxRadio, function (_ref4) {
  var theme = _ref4.theme;
  return theme.colors.keyFocus;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.colors.keyAccent;
}, FauxRadio, function (_ref6) {
  var theme = _ref6.theme;
  return theme.colors.text1;
}, FauxRadio, function (_ref7) {
  var theme = _ref7.theme;
  return theme.colors.ui1;
}, function (_ref8) {
  var theme = _ref8.theme;
  return theme.colors.ui1;
});
