// @looker/components/lib/Form/Fields/FieldText/FieldText.js
import "core-js/modules/es.object.assign.js";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { useID } from '../../../utils';
import { useFormContext } from '../../Form';
import { InputText } from '../../Inputs/InputText/InputText';
import { Field, omitFieldProps, pickFieldProps } from '../Field';
var FieldTextComponent = forwardRef(function (props, ref) {
  var id = useID(props.id);
  var validationMessage = useFormContext(props);
  return React.createElement(Field, _extends({
    id: id,
    validationMessage: validationMessage
  }, pickFieldProps(props)), React.createElement(InputText, _extends({}, omitFieldProps(props), {
    id: id,
    "aria-describedby": "describedby-".concat(id),
    validationType: validationMessage && validationMessage.type,
    ref: ref
  })));
});
FieldTextComponent.displayName = 'FieldTextComponent';
export var FieldText = styled(FieldTextComponent).withConfig({
  displayName: "FieldText",
  componentId: "rg64oz-0"
})([""]);
