// @looker/components/lib/cjs/Tooltip/useTooltip.js
"use strict";

require("core-js/modules/es.symbol.description.js");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.array.slice.js");

require("core-js/modules/es.array.from.js");

require("core-js/modules/es.weak-map.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTooltip = useTooltip;

var _react = _interopRequireWildcard(require("react"));

var _utils = require("../utils");

var _Portal = require("../Portal");

var _TooltipContent = require("./TooltipContent");

var _TooltipSurface = require("./TooltipSurface");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function useTooltip({
  canClose,
  content,
  isOpen: initializeOpen = false,
  width,
  maxWidth = '30rem',
  textAlign,
  disabled,
  id,
  invert,
  triggerElement,
  placement: propsPlacement = 'bottom',
  delay = 'intricate'
}) {
  const _useState = (0, _react.useState)(initializeOpen),
        _useState2 = _slicedToArray(_useState, 2),
        isOpen = _useState2[0],
        setIsOpen = _useState2[1];

  const _useAnimationState = (0, _utils.useAnimationState)(isOpen, delay, 'none'),
        busy = _useAnimationState.busy,
        className = _useAnimationState.className,
        renderDOM = _useAnimationState.renderDOM;

  const _useCallbackRef = (0, _utils.useCallbackRef)(),
        _useCallbackRef2 = _slicedToArray(_useCallbackRef, 2),
        surfaceElement = _useCallbackRef2[0],
        surfaceCallbackRef = _useCallbackRef2[1];

  const _useCallbackRef3 = (0, _utils.useCallbackRef)(),
        _useCallbackRef4 = _slicedToArray(_useCallbackRef3, 2),
        newTriggerElement = _useCallbackRef4[0],
        callbackRef = _useCallbackRef4[1];

  const element = typeof triggerElement === 'undefined' ? newTriggerElement : triggerElement;

  const handleOpen = () => {
    if (!disabled && (!element || !element.dataset.notooltip)) {
      setIsOpen(true);
    }
  };

  const handleClose = () => {
    if (canClose && !canClose()) return;
    setIsOpen(false);
  };

  const handleMouseOut = event => {
    if (!isOpen) return;
    const related = event.relatedTarget;

    if (element && (element === related || element.contains(related))) {
      return;
    }

    if (surfaceElement && (surfaceElement === related || surfaceElement.contains(related))) {
      return;
    }

    window.requestAnimationFrame(() => {
      handleClose();
    });
  };

  const usePopperProps = (0, _react.useMemo)(() => ({
    anchor: element,
    options: {
      modifiers: [{
        enabled: true,
        name: 'flip',
        options: {
          flipVariations: true
        }
      }],
      placement: propsPlacement
    }
  }), [element, propsPlacement]);

  const _usePopper = (0, _utils.usePopper)(usePopperProps),
        placement = _usePopper.placement,
        popperInstanceRef = _usePopper.popperInstanceRef,
        style = _usePopper.style,
        targetRef = _usePopper.targetRef;

  const ref = (0, _utils.useForkedRef)(targetRef, surfaceCallbackRef);
  const guaranteedId = (0, _utils.useID)(id);
  const popper = renderDOM && content && !disabled ? _react.default.createElement(_Portal.Portal, null, _react.default.createElement(_TooltipSurface.TooltipSurface, {
    "aria-busy": busy ? true : undefined,
    className: className,
    eventHandlers: {
      onMouseOut: handleMouseOut
    },
    placement: placement,
    ref: ref,
    style: style,
    maxWidth: maxWidth,
    invert: invert
  }, _react.default.createElement(_TooltipContent.TooltipContent, {
    role: "tooltip",
    id: guaranteedId,
    width: width,
    textAlign: textAlign
  }, content))) : null;
  return {
    domProps: {
      'aria-describedby': guaranteedId,
      className: renderDOM ? 'hover' : '',
      onBlur: handleClose,
      onFocus: handleOpen,
      onMouseOut: handleMouseOut,
      onMouseOver: handleOpen,
      ref: callbackRef
    },
    popperInstanceRef,
    tooltip: popper
  };
}
