// @looker/components/lib/cjs/UnorderedList/UnorderedList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnorderedList = void 0;

var _designTokens = require("@looker/design-tokens");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const typeVariant = (0, _styledSystem.variant)({
  prop: 'type',
  variants: {
    bullet: {
      listStyleType: 'disc',
      pl: 'medium'
    },
    none: {
      listStyleType: 'none'
    }
  }
});

const UnorderedList = _styledComponents.default.ul.withConfig({
  shouldForwardProp: _designTokens.shouldForwardProp
}).attrs(({
  color = 'body'
}) => ({
  color
})).withConfig({
  displayName: "UnorderedList",
  componentId: "qnps5j-0"
})(["", " ", " ", " ", " ", " ", " ", " li{margin-bottom:", ";}"], _designTokens.reset, _designTokens.textColor, _designTokens.typography, typeVariant, _designTokens.space, _designTokens.position, _designTokens.layout, ({
  theme
}) => theme.space.xxsmall);

exports.UnorderedList = UnorderedList;
