// @looker/components/lib/cjs/Portal/Portal.js
"use strict";

require("core-js/modules/es.weak-map.js");

require("core-js/modules/es.array.iterator.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Portal = exports.getPortalRoot = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactDom = require("react-dom");

var _styledComponents = _interopRequireDefault(require("styled-components"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const rootId = 'modal-root';

const getPortalRoot = () => {
  const existing = document.getElementById(rootId);

  if (existing) {
    return existing;
  } else {
    const newElement = document.createElement('div');
    newElement.id = 'modal-root';
    document.body.appendChild(newElement);
    return newElement;
  }
};

exports.getPortalRoot = getPortalRoot;
const Portal = (0, _react.forwardRef)((props, ref) => {
  const el = (0, _react.useRef)(document.createElement('div'));
  (0, _react.useEffect)(() => {
    const root = getPortalRoot();
    if (!root) return;
    const elCurrent = el.current;
    root.appendChild(elCurrent);
    return () => {
      root.removeChild(elCurrent);
    };
  }, [el]);

  const content = _react.default.createElement(InvisiBox, _extends({
    ref: ref
  }, props));

  return (0, _reactDom.createPortal)(content, el.current);
});
exports.Portal = Portal;
Portal.displayName = 'Portal';

const InvisiBox = _styledComponents.default.div.withConfig({
  displayName: "Portal__InvisiBox",
  componentId: "sc-8jnv99-0"
})(["align-items:", ";bottom:0;display:flex;justify-content:", ";left:0;pointer-events:none;position:fixed;right:0;top:0;z-index:", ";> *{pointer-events:auto;}"], ({
  vertical
}) => vertical === 'top' ? 'flex-start' : vertical === 'bottom' ? 'flex-end' : 'center', ({
  horizontal
}) => horizontal === 'left' ? 'flex-start' : horizontal === 'right' ? 'flex-end' : 'center', ({
  theme: {
    zIndexFloor
  }
}) => zIndexFloor);
