// @looker/design-tokens/lib/cjs/utils/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.disabledSwatchColor = exports.toggleSwitchShadowColor = exports.knobShadowColor = exports.calendarMixColor = exports.tabShadowColor = exports.iconButtonColorDerivation = exports.buttonShadow = void 0;

var _rgba = _interopRequireDefault(require("polished/lib/color/rgba"));

var _lighten = _interopRequireDefault(require("polished/lib/color/lighten"));

var _mix = _interopRequireDefault(require("polished/lib/color/mix"));

var _styledComponents = require("styled-components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const buttonShadow = (color = 'key') => (0, _styledComponents.css)(["box-shadow:0 0 0 0.15rem ", ";"], ({
  theme
}) => (0, _rgba.default)(theme.colors[color], 0.25));

exports.buttonShadow = buttonShadow;

const iconButtonColorDerivation = () => (0, _styledComponents.css)(["color:", ";"], ({
  theme
}) => (0, _lighten.default)(0.14, theme.colors.neutral));

exports.iconButtonColorDerivation = iconButtonColorDerivation;

const tabShadowColor = () => (0, _styledComponents.css)(["box-shadow:0 0 0 0.15rem ", ";"], ({
  theme
}) => (0, _rgba.default)(theme.colors.keyFocus, 0.25));

exports.tabShadowColor = tabShadowColor;

const calendarMixColor = () => (0, _styledComponents.css)(["color:", ";"], ({
  theme: {
    colors
  }
}) => (0, _mix.default)(0.65, colors.keyAccent, colors.neutralInteractive));

exports.calendarMixColor = calendarMixColor;

const knobShadowColor = () => (0, _styledComponents.css)(["box-shadow:0 0 0.01rem 0.01rem ", ";"], ({
  theme
}) => (0, _rgba.default)(theme.colors.keyInteractive, 0.5));

exports.knobShadowColor = knobShadowColor;

const toggleSwitchShadowColor = () => (0, _styledComponents.css)(["box-shadow:0 0 0 0.2rem ", ";"], ({
  theme
}) => (0, _rgba.default)(theme.colors.keyInteractive, 0.4));

exports.toggleSwitchShadowColor = toggleSwitchShadowColor;

const disabledSwatchColor = color => color && color !== 'transparent' ? (0, _rgba.default)(color, 0.85) : undefined;

exports.disabledSwatchColor = disabledSwatchColor;
