// @looker/design-tokens/lib/cjs/color/fallbacks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fallbackStateful = exports.fallbackBlends = void 0;

var _palette = require("../legacy/palette");

var _utils = require("./utils");

const fallbackBlends = {
  ui1: _palette.charcoal100,
  ui2: _palette.charcoal200,
  ui3: _palette.charcoal300,
  ui4: _palette.charcoal400,
  ui5: _palette.charcoal800,
  text1: _palette.charcoal400,
  text2: _palette.charcoal500,
  text3: _palette.charcoal600,
  text4: _palette.charcoal700,
  text5: _palette.charcoal800
};
exports.fallbackBlends = fallbackBlends;
const fallbackStateful = {
  keySubtle: _palette.purple000,
  keyAccent: _palette.purple100,
  keyFocus: _palette.purple300,
  keyInteractive: (0, _utils.generateInteractive)(_palette.purple400),
  keyPressed: (0, _utils.generatePressed)(_palette.purple400),
  keyText: _palette.white,
  keyBorder: _palette.purple400,
  criticalSubtle: _palette.red000,
  criticalAccent: _palette.red100,
  criticalFocus: _palette.red300,
  criticalInteractive: (0, _utils.generateInteractive)(_palette.red500),
  criticalPressed: (0, _utils.generatePressed)(_palette.red500),
  criticalText: _palette.white,
  criticalBorder: _palette.red400,
  neutralSubtle: _palette.charcoal000,
  neutralAccent: _palette.charcoal100,
  neutralFocus: _palette.charcoal300,
  neutralInteractive: (0, _utils.generateInteractive)(_palette.charcoal500),
  neutralPressed: (0, _utils.generatePressed)(_palette.charcoal500),
  neutralText: _palette.white,
  neutralBorder: _palette.charcoal400
};
exports.fallbackStateful = fallbackStateful;
