// @looker/components/lib/Dialog/Layout/DialogFooter.js
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.array.index-of.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.object.assign.js";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import styled from 'styled-components';
import { Space } from '../../Layout/Space';

var DialogFooterLayout = function DialogFooterLayout(_ref) {
  var children = _ref.children,
      secondary = _ref.secondary,
      props = _objectWithoutProperties(_ref, ["children", "secondary"]);

  return React.createElement(Space, _extends({
    as: "footer",
    reverse: true,
    between: true
  }, props), React.createElement(Space, {
    reverse: true
  }, children), secondary && React.createElement(Space, null, secondary));
};

export var DialogFooter = styled(DialogFooterLayout).attrs(function (_ref2) {
  var _ref2$px = _ref2.px,
      px = _ref2$px === void 0 ? 'xlarge' : _ref2$px,
      _ref2$py = _ref2.py,
      py = _ref2$py === void 0 ? 'large' : _ref2$py;
  return {
    px: px,
    py: py
  };
}).withConfig({
  displayName: "DialogFooter",
  componentId: "v5gtqt-0"
})(["flex-shrink:0;"]);
