// @looker/components/lib/Dialog/Layout/DialogLayout.js
import React from 'react';
import { Spinner } from '../../Spinner';
import { DialogContent } from './DialogContent';
import { DialogFooter } from './DialogFooter';
import { DialogHeader } from './DialogHeader';
export var DialogLayout = function DialogLayout(_ref) {
  var children = _ref.children,
      footer = _ref.footer,
      footerSecondary = _ref.footerSecondary,
      header = _ref.header,
      _ref$headerCloseButto = _ref.headerCloseButton,
      headerCloseButton = _ref$headerCloseButto === void 0 ? !footer && true : _ref$headerCloseButto,
      headerDetail = _ref.headerDetail,
      isLoading = _ref.isLoading;
  var dialogHeader = header && React.createElement(DialogHeader, {
    hideClose: !headerCloseButton,
    detail: headerDetail
  }, header);
  var dialogFooter = (footer || footerSecondary) && React.createElement(DialogFooter, {
    secondary: footerSecondary
  }, footer);
  return React.createElement(React.Fragment, null, dialogHeader, React.createElement(DialogContent, {
    hasFooter: !dialogFooter,
    hasHeader: !dialogHeader
  }, isLoading ? React.createElement(DialogLoading, null) : children), dialogFooter);
};

var DialogLoading = function DialogLoading() {
  return React.createElement(Spinner, {
    mx: "auto",
    my: "xxlarge"
  });
};
