// @looker/components/lib/Form/Fields/FieldTimeSelect/FieldTimeSelect.js
import "core-js/modules/es.object.assign.js";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { useID } from '../../../utils';
import { useFormContext } from '../../Form';
import { InputTimeSelect } from '../../Inputs/InputTimeSelect';
import { Field, omitFieldProps, pickFieldProps } from '../Field';
var FieldTimeSelectComponent = forwardRef(function (props, ref) {
  var validationMessage = useFormContext(props);
  var id = useID(props.id);
  return React.createElement(Field, _extends({
    "data-testid": "FieldSelectMultiId"
  }, pickFieldProps(props), {
    id: id,
    validationMessage: validationMessage
  }), React.createElement(InputTimeSelect, _extends({}, omitFieldProps(props), {
    "aria-labelledby": "labelledby-".concat(id),
    id: id,
    validationType: validationMessage && validationMessage.type,
    ref: ref
  })));
});
FieldTimeSelectComponent.displayName = 'FieldTimeSelectComponent';
export var FieldTimeSelect = styled(FieldTimeSelectComponent).withConfig({
  displayName: "FieldTimeSelect",
  componentId: "sc-18cfbnj-0"
})([""]);
