// @looker/components/lib/cjs/Dialog/Layout/DialogLayout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DialogLayout = void 0;

var _react = _interopRequireDefault(require("react"));

var _Spinner = require("../../Spinner");

var _DialogContent = require("./DialogContent");

var _DialogFooter = require("./DialogFooter");

var _DialogHeader = require("./DialogHeader");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DialogLayout = ({
  children,
  footer,
  footerSecondary,
  header,
  headerCloseButton = !footer && true,
  headerDetail,
  isLoading
}) => {
  const dialogHeader = header && _react.default.createElement(_DialogHeader.DialogHeader, {
    hideClose: !headerCloseButton,
    detail: headerDetail
  }, header);

  const dialogFooter = (footer || footerSecondary) && _react.default.createElement(_DialogFooter.DialogFooter, {
    secondary: footerSecondary
  }, footer);

  return _react.default.createElement(_react.default.Fragment, null, dialogHeader, _react.default.createElement(_DialogContent.DialogContent, {
    hasFooter: !dialogFooter,
    hasHeader: !dialogHeader
  }, isLoading ? _react.default.createElement(DialogLoading, null) : children), dialogFooter);
};

exports.DialogLayout = DialogLayout;

const DialogLoading = () => _react.default.createElement(_Spinner.Spinner, {
  mx: "auto",
  my: "xxlarge"
});
