// @looker/components/lib/cjs/Drawer/DrawerSurface.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DrawerSurface = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

var _SurfaceBase = require("../Dialog/SurfaceBase");

var _dialogWidth = require("../Dialog/dialogWidth");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const placement = (0, _styledSystem.variant)({
  prop: 'placement',
  variants: {
    left: {
      boxShadow: '-18px 0 18px -18px rgba(0, 0, 0, 0.12)',
      left: 0
    },
    right: {
      boxShadow: '-18px 0 18px -18px rgba(0, 0, 0, 0.12)',
      right: 0
    }
  }
});
const DrawerSurface = (0, _styledComponents.default)(_SurfaceBase.SurfaceBase).attrs(({
  placement = 'right',
  width = 'small'
}) => ({
  placement,
  width
})).withConfig({
  displayName: "DrawerSurface",
  componentId: "sc-1vwnqjh-0"
})(["height:100%;position:absolute;transition:transform ", ",opacity ", ";", " ", " &.entering,&.exiting{opacity:0.01;transform:translateX( ", " );}"], _SurfaceBase.surfaceTransition, _SurfaceBase.surfaceTransition, placement, _dialogWidth.dialogWidth, ({
  placement
}) => placement === 'left' ? '-100%' : '100%');
exports.DrawerSurface = DrawerSurface;
