// @looker/components/lib/cjs/List/ListItemLabel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListItemLabel = exports.ListItemLabelDiv = exports.ListItemLabelLink = exports.ListItemLabelButton = void 0;

require("core-js/modules/es.array.index-of.js");

var _designTokens = require("@looker/design-tokens");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

const ListItemLabelButton = _styledComponents.default.button.withConfig({
  displayName: "ListItemLabel__ListItemLabelButton",
  componentId: "nsrwox-0"
})([""]);

exports.ListItemLabelButton = ListItemLabelButton;

const ListItemLabelLink = _styledComponents.default.a.withConfig({
  displayName: "ListItemLabel__ListItemLabelLink",
  componentId: "nsrwox-1"
})([""]);

exports.ListItemLabelLink = ListItemLabelLink;

const ListItemLabelDiv = _styledComponents.default.div.withConfig({
  displayName: "ListItemLabel__ListItemLabelDiv",
  componentId: "nsrwox-2"
})([""]);

exports.ListItemLabelDiv = ListItemLabelDiv;

const listItemLabelElement = (itemRole, disabled) => {
  if (!disabled && itemRole === 'link') return ListItemLabelLink;
  if (itemRole === 'none') return ListItemLabelDiv;
  return ListItemLabelButton;
};

const ListItemLabelLayout = (_ref) => {
  let children = _ref.children,
      disabled = _ref.disabled,
      _ref$itemRole = _ref.itemRole,
      itemRole = _ref$itemRole === void 0 ? 'button' : _ref$itemRole,
      props = _objectWithoutProperties(_ref, ["children", "disabled", "itemRole"]);

  const Component = listItemLabelElement(itemRole, disabled);
  return _react.default.createElement(Component, _extends({
    disabled: disabled,
    type: itemRole === 'button' || disabled ? 'button' : undefined
  }, props), children);
};

const ListItemLabel = (0, _styledComponents.default)(ListItemLabelLayout).withConfig({
  shouldForwardProp: prop => prop === 'itemRole' || (0, _designTokens.shouldForwardProp)(prop)
}).withConfig({
  displayName: "ListItemLabel",
  componentId: "nsrwox-3"
})(["", " ", " align-items:center;border:none;color:inherit;cursor:pointer;display:flex;flex:1;font-size:inherit;font-weight:inherit;margin:0;min-width:0;outline:none;text-align:left;text-decoration:none;transition:", ";width:100%;&:hover,&:focus{color:inherit;text-decoration:none;}"], ({
  height,
  itemRole
}) => itemRole === 'none' && `height: ${height}px;`, _utils.listItemBackgroundColor, ({
  theme: {
    easings,
    transitions
  }
}) => `background ${transitions.quick}ms ${easings.ease},
  color ${transitions.quick}ms ${easings.ease}`);
exports.ListItemLabel = ListItemLabel;
