// @looker/components/lib/Form/Fields/FieldTime/FieldTime.js
import "core-js/modules/es.object.assign.js";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { useID } from '../../../utils';
import { useFormContext } from '../../Form';
import { InputTime } from '../../Inputs/InputTime';
import { Field, omitFieldProps, pickFieldProps } from '../Field';
var FieldTimeComponent = forwardRef(function (props, ref) {
  var validationMessage = useFormContext(props);
  var id = useID(props.id);
  return React.createElement(Field, _extends({}, pickFieldProps(props), {
    id: id,
    validationMessage: validationMessage
  }), React.createElement(InputTime, _extends({}, omitFieldProps(props), {
    "aria-describedby": "describedby-".concat(id),
    "aria-labelledby": "labelledby-".concat(id),
    id: id,
    validationType: validationMessage && validationMessage.type,
    ref: ref
  })));
});
FieldTimeComponent.displayName = 'FieldTimeComponent';
export var FieldTime = styled(FieldTimeComponent).withConfig({
  displayName: "FieldTime",
  componentId: "sc-105s2jn-0"
})(["width:100%;"]);
