// @looker/components/lib/OrderedList/OrderedList.js
import { reset, space, shouldForwardProp, textColor, typography, position, layout } from '@looker/design-tokens';
import styled from 'styled-components';
import { variant } from 'styled-system';
var typeVariant = variant({
  prop: 'type',
  variants: {
    letter: {
      listStyleType: 'upper-alpha',
      pl: 'medium'
    },
    none: {
      listStyleType: 'none'
    },
    number: {
      listStyleType: 'decimal',
      pl: 'medium'
    }
  }
});
export var OrderedList = styled.div.withConfig({
  shouldForwardProp: shouldForwardProp
}).attrs(function (_ref) {
  var _ref$color = _ref.color,
      color = _ref$color === void 0 ? 'body' : _ref$color,
      _ref$type = _ref.type,
      type = _ref$type === void 0 ? 'none' : _ref$type;
  return {
    as: 'ol',
    color: color,
    type: type
  };
}).withConfig({
  displayName: "OrderedList",
  componentId: "csldlu-0"
})(["", " ", " ", " ", " ", " ", " ", " li{margin-bottom:", ";}"], reset, textColor, typography, typeVariant, space, position, layout, function (_ref2) {
  var theme = _ref2.theme;
  return theme.space.xxsmall;
});
